src/Agent.Worker/StepsRunner.cs

```diff
@@ -205,5 +204,36 @@ public sealed class StepsRunner : AgentService, IStepsRunner
                 Trace.Info($"Current state: job state = '{jobContext.Result}', step failed = {stepFailed}, critical step failed = {criticalStepFailed}");
             }
         }
+
+        private bool ShouldRunStep(IStep step, bool stepFailed, bool criticalStepFailed)
+        {            
+            switch(step.RunMode)
+            {
+                // The server hasn't been upgraded yet. Fallback to old code
+                case RunMode.Undefined:
```

Comment:
yes the agent is compatible back to TFS 2015 on Linux and back to TFS 2017 RTM on Windows. but can the properties be deleted from IStep and infer RunMode based on the job message?